database/src/ops/block.rs

```diff
@@ -1,89 +1,398 @@
+    tx_rw: &mut Rw,
+    height_and_blocks: &[(BlockHeight, BlockInfoLatest)],
+) -> Result<(), RuntimeError>
+where
+    Ro: TxRo<'env>,
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    add_blocks_v3(env, tx_rw, height_and_blocks)
 }
 
 /// TODO
-pub fn pop_block() {
-    todo!()
+///
+/// # Errors
+/// TODO
+#[inline]
+pub fn add_block_v1<'env, Ro, Rw, Env>(
+    env: &Env,
+    tx_rw: &mut Rw,
+    height: BlockHeight,
+    block: &BlockInfoV1,
+) -> Result<(), RuntimeError>
+where
+    Ro: TxRo<'env>,
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    env.open_db_rw::<BlockInfoV1s>(tx_rw)?.put(&height, block)
 }
 
 /// TODO
-pub fn block_exists() {
-    todo!()
+///
+/// # Errors
+/// TODO
+#[inline]
+pub fn add_blocks_v1<'env, Ro, Rw, Env>(
+    env: &Env,
+    tx_rw: &mut Rw,
+    height_and_blocks: &[(BlockHeight, BlockInfoV1)],
+) -> Result<(), RuntimeError>
+where
+    Ro: TxRo<'env>,
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    let mut table = env.open_db_rw::<BlockInfoV1s>(tx_rw)?;
+    for (height, block) in height_and_blocks {
+        table.put(height, block)?;
+    }
+    Ok(())
 }
 
 /// TODO
-pub fn get_block_hash() {
-    todo!()
+///
+/// # Errors
+/// TODO
+#[inline]
+pub fn add_block_v2<'env, Ro, Rw, Env>(
+    env: &Env,
+    tx_rw: &mut Rw,
+    height: BlockHeight,
+    block: &BlockInfoV2,
+) -> Result<(), RuntimeError>
+where
+    Ro: TxRo<'env>,
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    env.open_db_rw::<BlockInfoV2s>(tx_rw)?.put(&height, block)
 }
 
 /// TODO
-pub fn get_block_height() {
-    todo!()
+///
+/// # Errors
+/// TODO
+#[inline]
+pub fn add_blocks_v2<'env, Ro, Rw, Env>(
+    env: &Env,
+    tx_rw: &mut Rw,
+    height_and_blocks: &[(BlockHeight, BlockInfoV2)],
+) -> Result<(), RuntimeError>
+where
+    Ro: TxRo<'env>,
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    let mut table = env.open_db_rw::<BlockInfoV2s>(tx_rw)?;
+    for (height, block) in height_and_blocks {
+        table.put(height, block)?;
+    }
+    Ok(())
 }
 
 /// TODO
-pub fn get_block_weight() {
-    todo!()
+///
+/// # Errors
+/// TODO
+#[inline]
+pub fn add_block_v3<'env, Ro, Rw, Env>(
+    env: &Env,
+    tx_rw: &mut Rw,
+    height: BlockHeight,
+    block: &BlockInfoV3,
+) -> Result<(), RuntimeError>
+where
+    Ro: TxRo<'env>,
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    env.open_db_rw::<BlockInfoV3s>(tx_rw)?.put(&height, block)
 }
 
 /// TODO
-pub fn get_block_already_generated_coins() {
+///
+/// # Errors
+/// TODO
+#[inline]
+pub fn add_blocks_v3<'env, Ro, Rw, Env>(
+    env: &Env,
+    tx_rw: &mut Rw,
+    height_and_blocks: &[(BlockHeight, BlockInfoV3)],
+) -> Result<(), RuntimeError>
+where
+    Ro: TxRo<'env>,
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    let mut table = env.open_db_rw::<BlockInfoV3s>(tx_rw)?;
+    for (height, block) in height_and_blocks {
+        table.put(height, block)?;
+    }
+    Ok(())
+}
```

Comment:
These bulk functions are needed right?

Otherwise calling `get_block()` in a loop would mean opening+closing the table in a loop, where as in this function it's done only once.

Reply:
yeah I would say so but I don't think `add_blocks_v3` needs to be a bulk function just `add_blocks`

Reply:
So just `add_block()` and `add_blocks()` is needed, both which branch and add to the appropriate table depending on the hard fork?

Reply:
yep, also I think `add_block` should do more than just add to the block info table, it should fill in everything for that block. Each tx/ output etc.